### ISSUE TITLE
RNNTloss decrease while the syllabel error rate increace

### ISSUE THREAD

**Issue by liangzhenlin123** [HawkAaron/warp-transducer#27]
Hello, I used RNNT training on the Chinese speech recognition library of more than 300 hours (the encoder did pretrain, but the decoder is a random initialization parameter). After training dozens of epoch, the loss first quickly dropped from more than 1000 to 60. Then slowly dropped to more than 20, but the SER of inference has risen from 2 to 20. Is this normal? It seems that you mentioned this phenomenon elsewhere.
Thank you very much!

**Comment by HawkAaron** [comment]
Did you mean the validation loss? How did you calculate the SER after each epoch?

BTW, since you mentioned the encoder was pretrained, then whats the performance of this encoder?

**Comment by liangzhenlin123** [comment]
Hi, I am doing the inference of the SER on the training set. Every time I train a batch, I do inference on the batch, and then calculate the editing distance from the actual syllable.






At 2019-06-26 15:41:24, "Mingkun Huang" <notifications@github.com> wrote:


Did you mean the validation loss? How did you calculate the SER after each epoch?

BTW, since you mentioned the encoder was pretrained, then whats the performance of this encoder?

**Comment by liangzhenlin123** [comment]
Hi, I am doing the inference of the SER on the training set. Every time I train a batch, I do inference on the batch, and then calculate the editing distance from the actual syllable. When both the prediction model and the ctc encoder are pre-trained, the loss is slower, but the SER is still rising. When the loss drops to around 2, the SER oscillates around 10.
The RNNTloss looks like:
and SER:
Thanks a lot!





At 2019-06-26 15:41:24, "Mingkun Huang" <notifications@github.com> wrote:


Did you mean the validation loss? How did you calculate the SER after each epoch?

BTW, since you mentioned the encoder was pretrained, then whats the performance of this encoder?

**Comment by HawkAaron** [comment]
Generally, we test the model when it converged on the validation set. Could you please show the full train & valid loss curve? Maybe you'd better calculate SER on the validation set to select the best model.

**Comment by liangzhenlin123** [comment]
> Generally, we test the model when it converged on the validation set. Could you please show the full train & valid loss curve? Maybe you'd better calculate SER on the validation set to select the best model.

Thanks for your reply, however, i think the loss on the training set drops but the SER rises very strangely, The validation set may go up, but why does the training set rise? The figure likes:
RNNTloss:
![image](https://user-images.githubusercontent.com/29907023/60229429-6f517000-98c8-11e9-867f-59f35db9a13e.png)
SER on train set (batch):
![image](https://user-images.githubusercontent.com/29907023/60229457-8db76b80-98c8-11e9-9d0d-7efefa3c37f2.png)

Though i hope the SER would be lower again, the dataset has been trained dozens epochs ：)

**Comment by liangzhenlin123** [comment]
The SER i mentioned here is the edit distance of syllabel seqence

**Comment by HawkAaron** [comment]
Have you got SER on the validation set?

**Comment by liangzhenlin123** [comment]
> Have you got SER on the validation set?

Not yet, you mean although it works bad on train data set, the model may get a better SER?  BTW, I wonder why the SER of the training set will rise, and I have used the pre-trained CE model and the ctc model as the initialization parameters of the RNNT. Although the RNNTloss drops faster, the SER will still rise. I am sure my inference is no problem because i have compared the predicted syllable sequence with the actual print. The RNNT predicted syllable sequence has same length with the input wav frames (do a loop in ctc's output when inference), and after removing blanks and merging same labels, there still has a lot of wrong pred labels inserted in two right labels.  
How did you solve this problem? 
your issue:  https://github.com/awni/speech/issues/27

**Comment by HawkAaron** [comment]
The RNN-T decoding algorithm is different with CTC,  please refer to my [RNN-Transducer](https://github.com/HawkAaron/RNN-Transducer/blob/graves2013/model.py#L81) repo for more details.

**Comment by liangzhenlin123** [comment]
> The RNN-T decoding algorithm is different with CTC, please refer to my [RNN-Transducer](https://github.com/HawkAaron/RNN-Transducer/blob/graves2013/model.py#L81) repo for more details.

I didn't use the ctc decoder. I just rewrote the version of tensorflow with reference to the decoder you wrote.

**Comment by HawkAaron** [comment]
@liangzhenlin123 You'd better test the model after it been fully converged. And please show the full validation loss & SER if you got.

**Comment by FieldRen** [comment]
I have a similar problem. 
I am trying to use Transducer in **OCR task**.
My implement details as follows:

1.  I also used RNN-T tensorflow binding to train my model，and rewrote the greedy decode of tenorflow version  refering to your [RNN-Teansducer](https://github.com/HawkAaron/RNN-Transducer/blob/graves2013/model.py#L81);

2. I used pretrained CTC model(loss=±8)(accuracy = 0.94 at validation dataset) as encoder, the decoder is initialized by random parameter. The RNN-Transducer loss drops quickly from 400 to 5，but validation accuracy is very bad, accuracy = 0.65 at validation dataset. when I continue to train the Transducer, the RNN-Transducer loss drops to 2, meanwhile the validation accuracy is also drops and becomes worse with result of validation accuracy =0.26

3. I also tried training Transducer from scratch, the loss can droped to 1, but the validation accuracy is just 0.63 (validation accuracy of CTC is 0.94). I am confused about this. Why loss is so low, but accuracy is bad.  In order to locate the problem, I just trained one data, and the loss drops to 0, then I do inference using this train data, the decoded result is accuracy=0.7. My loss is 0, inference result also have error.  In my impression, when loss is 0, the inference result should Completely correct, the inference data is just the trained data!

So, I have some questions here：

I think the problem is caused by decode method.

Maybe, my greedy decode impematation is wrong. Because it is hard to write loop and condition in Tensorflow, I am trying to use TF2.0. BTW, could you tell me have you implemented the greedy decode of TF?

**Comment by FieldRen** [comment]
hello @liangzhenlin123， could you tell me your emails, I have something about tensorflow RNNT to consult. 
**my emails**: fieeldren@gmail.com 
Looking forward for your reply.
Thank you very much!

**Comment by FieldRen** [comment]
Hi~ @liangzhenlin123 , has your problem been solved? I have the same problem. With the decreasing of the RNNT loss, the predict result in training set becomes worse. I used CTC pretrained , and implement greedy decode of tensorflow. Can we talk about it?

thanks a lot!

**Comment by liangzhenlin123** [comment]
> Hi~ @liangzhenlin123 , has your problem been solved? I have the same problem. With the decreasing of the RNNT loss, the predict result in training set becomes worse. I used CTC pretrained , and implemented greedy decode of tensorflow. Can we talk about it?
> 
> Thanks a lot!

I solve it by initializing the prediction part with a CE pretrained RNNLM, u can have a try:)

**Comment by FieldRen** [comment]
> I solve it by initializing the prediction part with a CE pretrained RNNLM, u can have a try:)

Thank you very much!
